Bugfixes

MathieuBordere opened the pull request:
- There was an issue with sending AppendEntries RPC's containing entries while a snapshot was being installed by a follower. Now, we will always send empty AppendEntries RPC's to a follower that is installing a snapshot. This should fix most of @morphis his scaling issues together with https://github.com/lxc/lxd/pull/8629. 

- Protect against an InstallSnapshot RPC that arrives while entries are being saved to disk and could lead to inconsistent state.

- Don't commit entries when handling a HeartBeat while we are installing a snapshot.

stgraber commented:
Looks reasonable to me, will give a bit of time for @freeekanayaka to take a look though as I'm definitely not familiar enough with this code ;)

freeekanayaka commented:
How hard would it be to have unit tests that cover one on more of the mentioned edge cases? (i.e. send AppendEntries RPCs while installing a snapshot, ignore snapshots while persisting entries, and skip commit when installing a snapshot).

MathieuBordere commented:
> How hard would it be to have unit tests that cover one on more of the mentioned edge cases? (i.e. send AppendEntries RPCs while installing a snapshot, ignore snapshots while persisting entries, and skip commit when installing a snapshot).

You're right, I should test them.

freeekanayaka commented:
> > How hard would it be to have unit tests that cover one on more of the mentioned edge cases? (i.e. send AppendEntries RPCs while installing a snapshot, ignore snapshots while persisting entries, and skip commit when installing a snapshot).
> 
> You're right, I should test them.

I'm not sure how hard it will be to reproduce those scenarios, hopefully the tooling under `test/integration` are enough. If it proves to be too much effort, it might not be worth. YMMV

MathieuBordere commented:
> How hard would it be to have unit tests that cover one on more of the mentioned edge cases? (i.e. send AppendEntries RPCs while installing a snapshot, ignore snapshots while persisting entries, and skip commit when installing a snapshot).

- AppendEntries RPCs while installing a snapshot: This is already covered in a test, however I have a hard time reproducing in a unit test, the issue I encountered while running a reproducer script for an Anbox scaling problem.
- ignore snapshots while persisting entries: Added a test with an InstallSnapshot RCP arriving during persisting entries
- skip commit when installing a snapshot: This is already covered, tests started failing when I changed 
  `r->commit_index = min(args->leader_commit, logLastIndex(&r->log));` to 
  `r->commit_index = min(args->leader_commit, r->last_stored);` and are fixed by adding the check to skip committing while installing a snapshot.

freeekanayaka reviewed (approved):
Looks good to me.
